[torch][segment_reduce] Add support for multi-dimensional input (cuda)

serhaty opened the pull request:
Summary:
Same as title.

Next Steps:
- Add support for multi-d cuda backward
- Add support for sum/min
- More testing and benchmarking
- Cleanup
    - Update default values when length is 0
    - Use TensorIterator
    - Update documentation

Test Plan: Unit test to cover cuda forward path.

Differential Revision: D29135373

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -70,21 +71,192 @@ __global__ static void post_sum_div_kernel(
+    int64_t starting_index = (j * stride_count) + lane_id;
+    const auto data = values_data[starting_index];
+    // TODO: There is no need to branch with every element
+    if (reduction == SegmentReductionType::MAX) {
```
if perf is important, you can make `reduction` a template parameter, and branch will be eliminated at compile time

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -70,21 +71,192 @@ __global__ static void post_sum_div_kernel(
+      initial_value =
+          at::_isnan(data) ? data : std::max<scalar_t>(initial_value, data);
+    } else if (reduction == SegmentReductionType::MEAN) {
+      initial_value = at::_isnan(data) ? data : (initial_value + data);
```
no need for ternary operator here, `initial_value + data` will be `nan` if `data` is nan

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -70,21 +71,192 @@ __global__ static void post_sum_div_kernel(
+  if (lengths_data[row_id] == 0 && !is_initial_set) {
+    output_data[output_index] = static_cast<scalar_t>(NAN);
+  } else {
+    output_data[output_index] = initial_value;
```
avoid multiple global memory accesses (it would happen if lengths is not 0), instead compute this value in register (`scalar_t out_tmp = initial value;` and `out_tmp` in the conditional below) and write it out unconditionally. Compiler is usually not smart enough to optimize away all global reads and writes.

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -97,54 +269,78 @@ Tensor _segment_reduce_cuda_kernel(
+          initial_value = 0;
+        }
+
+        if (output_shape.size() > 1) {
```
if input/output is multi-d, but the non-reduced dimensions are 1 (input.size(0) == input.numel()) you are better off taking CUB path, but that's an optimization for later.

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -70,21 +71,192 @@ __global__ static void post_sum_div_kernel(
+        auto* grad_input_data = grad_input.data_ptr<scalar_t>();
+        const auto* values_data = data_contig.data_ptr<scalar_t>();
+
+        segment_reduce_backward_kernel<scalar_t>
```
this will be pretty inefficient for 1d inputs, but ok for now.

ngimel reviewed (commented):
This looks really good! I left minor comments, and then it will be good to go.

Commit:
Collapsed revision

Differential Revision:
D29051902
* [torch][segment_reduce] Add support for multi-d backward (cpu)

D29067989

D28922838 (https://github.com/pytorch/pytorch/commit/f9445c8a6b606dec9e59a8d8fd4e90a7d4e128d1)

D28922838 (https://github.com/pytorch/pytorch/commit/f9445c8a6b606dec9e59a8d8fd4e90a7d4e128d1)

D29051902

D29105457

D29105457
D29105457

fbshipit-source-id: 0fbb1b98e8861e4c7ac008bd0d41e973387a4aac

Commit:
[torch][segment_reduce] Add support for multi-dimensional input (cuda) (#60018)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/60018

Same as title. This diff finishes cuda support for currently implemented reductions and input parameters.

Next Steps:
- Add support for sum/min
- More testing and benchmarking
- Cleanup
    - Update default values when length is 0
    - Use TensorIterator
    - Update documentation

Test Plan: Unit test to cover cuda forward path.

Differential Revision: D29135373

fbshipit-source-id: ee6f5f9a122dbad2ca983c4efde2ac9b2097393d

ngimel reviewed (approved):
lgtm as long as tests pass
